ApolloGateway: Construct and use RemoteGraphQLDataSource to issue introspection query to Federated Services

mcohen75 opened the pull request:
This is meant to address #2891.

RemoteGraphQLDataSource provides a nice way to enhance requests before they are sent to federated services. This same mechanism is suitable for enhancing requests sent to federated services when issuing an enhanced introspection query.

`ApolloGateway` currently accepts an optional HeadersInit parameter that facilitates enhancing the request with headers. This is sufficient if all you need to do is send the same headers to all federated services. In many environments, however, services participating in federation will have diverse requirements. Note that this change is compatible with the HeadersInit parameter.

This change provides:
1. Access to the full range of request enhancement provided by `RemoteGraphQLDataSource`'s willSendRequest function. Notably, this allows enhancing the url, headers and extensions.
2. The ability to enhance requests on a per service basis.

A side benefit is a small savings in duplicated code as RemoteGraphQLDataSource already knows how to issue a GraphQL request over HTTP.

mcohen75 commented:
@JacksonKearl given your involvement in #2891 and the introduction of the `introspectionHeaders` parameter I think this change may be interesting to you.

mcohen75 commented:
Is this something that we could incorporate @JakeDawkins @zionts? I believe this provides very useful functionality and is a sane way to enrich introspection requests. If there is anything else I can do here to move this forward please do let me know.

JacksonKearl commented:
Hey @mcohen75, thanks for this! We're definitely interested in expanding the configurability of the introspection request. One concern I have with this as it stands is that it is a breaking change on the old API (by changing the parameters of `protected` methods). While we are still in pre-1.0 and thus could in theory adopt this with a minor, it would be nice to have backwards compatibility. @jbaxleyiii what do you think re. backwards compatibility?

mcohen75 commented:
Thank you for following up @JacksonKearl! I've updated my solution here to preserve backwards compatibility. The only downside is that I end up creating the remote data source just to issue the enhanced introspection query. This seems like a reasonable tradeoff to me if we'd like to preserve backwards compatibility. 

I chose to go this route since services are only added to the serviceMap after the schema is loaded and validated. I wasn't sure if this constraint is one that we wanted to violate.

trevor-scheer commented on packages/apollo-gateway/src/loadServicesFromRemoteEndpoint.ts:
```diff
@@ -22,31 +26,34 @@ export async function getServiceDefinitionsFromRemoteEndpoint({
   // for each service, fetch its introspection schema
   const services: ServiceDefinition[] = (await Promise.all(
-    serviceList.map(service => {
-      if (!service.url) {
```
Was it intentional to remove this? I don't believe removing is necessary in the scope of this PR, but please correct me if I'm wrong. Restoring this would also remove the need to cast url to `<string>` below, since we'd be guaranteeing the truthiness of `serviceDefinition.url`.

mcohen75 replied:
I removed this check here because the same check exists when creating the data source. Since the url is used in this method as well it makes sense to retain the check here as well. I'll restore this check here.

trevor-scheer commented on packages/apollo-gateway/src/loadServicesFromRemoteEndpoint.ts:
```diff
@@ -22,31 +26,34 @@ export async function getServiceDefinitionsFromRemoteEndpoint({
+
+      return dataSource
+        .process({ request, context: {} })
+        .then(({ data, errors }: GraphQLResponse) => {
```
Nit: TS should be able to infer the `GraphQLResponse` type based on the return type of `RemoteGraphQLDataSource.process()`.

trevor-scheer commented on packages/apollo-gateway/src/index.ts:
```diff
@@ -236,8 +242,13 @@ export class ApolloGateway implements GraphQLService {
 
     if (isRemoteConfig(config)) {
+      const serviceList = config.serviceList.map(serviceDefinition => ({
+        serviceDefinition,
```
Nit: I'd be interested in flattening this structure into `{ name, url, dataSource }`.

mcohen75 commented:
> add a few tests that demonstrate the successful addition of headers on a per-service basis
Will do!

mcohen75 commented on __mocks__/apollo-server-env.ts:
```diff
@@ -15,8 +15,8 @@ import {
 interface FetchMock extends jest.Mock<typeof fetch> {
-  mockResponseOnce(data?: any, headers?: HeadersInit, status?: number): void;
-  mockJSONResponseOnce(data?: object, headers?: HeadersInit): void;
+  mockResponseOnce(data?: any, headers?: HeadersInit, status?: number): this;
```
This change enables setting up multiple mock invocations. I believe `void` is technically an incorrect return type as the underlying API used here actually returns `this`. See https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/jest/index.d.ts#L997.

trevor-scheer replied:
👍

mcohen75 commented:
@trevor-scheer I have added tests to confirm that 1) enhanced introspection requests are made using the data source and 2) requests can be customized per service. I've also addressed the other feedback you provided.

trevor-scheer commented:
Tests LGTM, thank you for taking the time @mcohen75 🙇

zionts commented:
@trevor-scheer @mcohen75 I wonder if we might be able to use this functionality to enable some simple retry semantics on federation introspection as well, since services going down or experiencing brief interruptions to service shouldn't cause the gateway to crash on startup. I'm assuming that a `RemoteGraphQLDataSource` will not automatically issue retries due to 5xx errors, but including some basic timeouts and retries would likely go a long way :)

alex-hioperator commented:
Hey guys! Super excited this feature is getting added--its something we need to handle authentication on introspection since google cloud IAM requires getting an auth token for a specific URL for each query.

I'm a bit newer to the community here, so I was just wondering if there's an example of the new way to set the introspection query?

trevor-scheer commented:
@alex-hioperator, we do have some docs with an example. Let me know if you have any questions beyond what's covered here 👍

https://www.apollographql.com/docs/apollo-server/federation/implementing/#sharing-context-across-services

alex-hioperator commented:
@trevor-scheer thanks for the super fast reply! 

The bit I'm unclear on is how to use the flexibility of willSendRequest in that initial introspection query. Though looking at the docs now, I do see the recommendation of using the schema from the registry for production rather than introspection--so that would actually work. I'd just need to configure it as such.

cluedtke commented:
> The bit I'm unclear on is how to use the flexibility of willSendRequest in that initial introspection query.  

Agree with this ^^

Our federated endpoints are behind auth, so even introspection queries require an auth header. `introspectionHeaders` works, but user tokens (for non-introspection queries) are also sent to gateway and need to be forwarded to the federated services, which requires `willSendRequest()`. Implementing `willSendRequest()` causes `introspectionHeaders` not to work..

So we had to remove `introspectionHeaders` and use `willSendRequest()` exclusively. The smelly part of this is that now in `willSendRequest()` we have a conditional, which generates an introspection auth header **if** the `request.query` is GetServiceDefinition.

Pitazzo commented:
@cluedtke would you show me how do you generate those introspection auth headers?
